autoPyTorch/data/tabular_feature_validator.py

```diff
@@ -246,43 +267,51 @@ def _check_data(
         # Then for Pandas, we do not support Nan in categorical columns
         if hasattr(X, "iloc"):
             # If entered here, we have a pandas dataframe
-            X = typing.cast(pd.DataFrame, X)
+            X = cast(pd.DataFrame, X)
 
             # Handle objects if possible
-            if not X.select_dtypes(include='object').empty:
+            exist_object_columns = has_object_columns(X.dtypes.values)
+            if exist_object_columns:
                 X = self.infer_objects(X)
 
             # Define the column to be encoded here as the feature validator is fitted once
             # per estimator
-            enc_columns, _ = self._get_columns_to_encode(X)
-
+            # enc_columns, _ = self._get_columns_to_encode(X)
             column_order = [column for column in X.columns]
             if len(self.column_order) > 0:
                 if self.column_order != column_order:
                     raise ValueError("Changing the column order of the features after fit() is "
                                      "not supported. Fit() method was called with "
                                      "{} whereas the new features have {} as type".format(self.column_order,
-                                                                                          column_order,)
+                                                                                          column_order, )
                                      )
             else:
                 self.column_order = column_order
 
             dtypes = [dtype.name for dtype in X.dtypes]
             if len(self.dtypes) > 0:
-                if self.dtypes != dtypes:
-                    raise ValueError("Changing the dtype of the features after fit() is "
-                                     "not supported. Fit() method was called with "
-                                     "{} whereas the new features have {} as type".format(self.dtypes,
-                                                                                          dtypes,
-                                                                                          )
-                                     )
+                dtypes_diff = [s_dtype != dtype for s_dtype, dtype in zip(self.dtypes, dtypes)]
+                if any(dtypes_diff):
+                    if self.all_nan_columns is not None and len(self.all_nan_columns) > 0:
+                        if len(set(X.columns[dtypes_diff]).difference(self.all_nan_columns)) != 0:
```

Comment:
If I understand the logic correctly, this condition `self.all_nan_columns is not None and len(self.all_nan_columns) > 0` does not matter?
Because if we remove by this condition, we end up with allowing some cases where dtypes are different in the training and test.

For example, when we have column `A (int)`, `B (int)`  in training and `A (int)`, `B (float)` in test where `B` is not `all_nan` in both training and test.
Then we cannot filter such cases.

Reply:
I mean we may have a case of which `len(self.all_nan_columns) == 0` but we should remove.
Let me make sure my point.
First, there are three conditions:
`C1`. `any(dtypes_diff)`: At least one of the columns has different data type between training and test datasets
`C2`. `len(self.all_nan_columns) > 0`: There is at least one column in the dataset that only has nan in the column
`C3`. `len(set(X.columns[dtypes_diff]).difference(self.all_nan_columns)) != 0`: A set of columns that have different data type in training and test and a set of columns that have only nan in test are identical

We raise the error if and only `C1`, `C2` and `C3` are satisfied.
But I am saying the condition `C1 and C3` also should be failed.

Reply:
Yes, I got your point.
But I am saying your condition does not cover everything, does it?
Yours will miss the following case:
```
train_data = {'A': 1}
test_data = {'A': 1.0}
```